Бочаров Александр

Geratoptus opened the pull request.

Geratoptus committed:
Создал проект TagsCloudVisualization.csproj

Geratoptus committed:
Создал интерфейс для раскладки ICircularCloudLayouter.cs

Geratoptus committed:
Создал интерфейс для генератора точек спирали

Geratoptus committed:
Добавил точку входа

Geratoptus committed:
Сделал красный тест для конструктора FermatSpiralPointsGenerator.cs

Geratoptus committed:
Теперь конструктор FermatSpiralPointsGenerator проходит тест

Geratoptus committed:
Создал красный тест для метода GeneratePoints

Geratoptus committed:
Теперь тест для метода GeneratePoints зеленый

Geratoptus committed:
FermatSpiralPointsGenerator рефактор

Geratoptus committed:
Создал CircularCloudLayouter.cs

Geratoptus committed:
Создал CircularCloudLayouterTest.cs с красным тестом для метода PutNextRectangle

Geratoptus committed:
Теперь тест для метода PutNextRectangle зеленый

Geratoptus committed:
Теперь проект структурирован по папочкам

Geratoptus committed:
Теперь namespace'ы в соответствии с папочками

Geratoptus committed:
Поставил System.Drawing.Common для Bitmap

Geratoptus committed:
Сделал тестам правильные using'и namespace'ов

Geratoptus committed:
Поправил using'и в тестах

Geratoptus committed:
Добавил интерфейс для Visualizer'а

Geratoptus committed:
Добавил дефолтную реализацию DefaultVisualizer.cs

Geratoptus committed:
Теперь в Main'е производится раскладка с параметрами и затем сохранение визуализации получившегося

Geratoptus committed:
Добавил README.MD с картинками

Geratoptus committed:
README.MD теперь должен нормально отображаться

Geratoptus committed:
Теперь README.MD точно должен нормально отображаться

Geratoptus committed:
Забыл TagsCloudVisualizationTests закоммитить

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,48 @@
+
+[TestFixture]
+[TestOf(typeof(CircularCloudLayouter))]
+public class CircularCloudLayouterTest
```
Использовал tdd для написания тестов?

Текущий тест выглядит уже весьма комплексно.
Можно было бы начать с самых простых тестов, наподобие:
- прямоугольник устанавливается
- прямоугольник правильного размера
- и далее, далее, далее

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,48 @@
+    [Repeat(10)]
+    public void PutNextRectangle_ShouldReturnRectanglesWithoutIntersections()
+    {
+        var randomizer = new Random();
```
Не будет ли тут проблем с рандомом?
[Утверждается](https://habr.com/ru/articles/165459/), что подобная запись может приводить к генерации детерминированой последовательности.

Предлагаю два пути:
- Доказать, что такая запись будет приводить к генерации разных последовательностей в современном dotnet
- Исправить создание экземпляра `Random`

Folleach replied:
\+ потенциальная проблема, не уверен сам, нужно ли её исправлять, поэтому исправление на твоё усмотрение:
Сейчас тест генерирует случайные последовательности, что означает, что результат теста может отличаться от запуска к запуску. Если проблема будет повтоярться редко, кажется, будет сложно найти значения, на которых она происходит. Или нет и всё будет элементарно просто?

Folleach replied:
Думаю, хватило бы просто картинки.  
Под значениями, на которых тест ломается, я имел ввиду вводные данные, такие как параметры спирали и центра )
Но в любом случае этого не требовал, поэтому пусть будет так

resolved

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        IsIntersectionBetweenRectangles(rectangles).Should().BeFalse();
+    }
```
Несмотря на то, что тест написан с использованием AAA, сейчас его читать сложно, потому что он содержит в себе много ответственности.  
Выделишь вспомогательные методы, которые могут пригодиться в других тестах? Обрати внимание, что важно найти правильное место, куда будет вынесен тот или иной дополнительный метод :)

Folleach replied:
Стало удобнее читать )
Как по мне, больший вклад внесли операции генерации рандомного размера и точки, а также переписанный `PutRectanglesInLayouter` на Linq

resolved

Folleach commented on cs/TagsCloudVisualizationTests/FermatSpiralPointsGeneratorTest.cs:
```diff
@@ -0,0 +1,46 @@
+     TestCase(0, 1, Description = "Radius is zero"),
+     TestCase(1, 0, Description = "AngleOffset is zero"),
+     TestCase(1, -1, Description = "Negative angleOffset")]
+    public void ShouldThrowArgumentException_AfterWrongCreation(double radius, double angleOffset)
```
Интересное перечисление атрибутов - через запятую.
Да, так можно, но может быть пусть каждый будет обёрнут в свою пару `[]`? Так дублировать их будет легче :)

Folleach replied:
resolved, но будь внимательнее с пробелами )

Folleach commented on cs/TagsCloudVisualizationTests/FermatSpiralPointsGeneratorTest.cs:
```diff
@@ -0,0 +1,46 @@
+        new TestCaseData(2, 180, 1).Returns(new Point(-1, 0)),
+        new TestCaseData(4, 90, 1).Returns(new Point(0, 1)),
+        new TestCaseData(4, 300, 1).Returns(new Point(2, -3))
+    };
```
Ну, хорошо, понятно почему так сделано - Point не записать в атрибут.  

Разве что спрошу тут, какую версию языка ты используешь? В C# 12 появилась возможность создавать массивы альтернативным синтаксисом:
```cs
int[] array = [ 1, 2, 3 ];
```
https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#collection-expressions
Что, порой, бывает удобно

Folleach replied:
resolved

Круто, что применил в других местах тоже!

Folleach commented on cs/TagsCloudVisualization/PointsGenerators/FermatSpiralPointsGenerator.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        var radiusVector = OffsetPerRadian * angle;
+        
+        var x = (int)Math.Round(
+            radiusVector * Math.Cos(angle) + spiralCenter.X);
```
Зачем тут перенос?

Folleach replied:
Рассказал в личных сообщениях, почему так, ну, хорошо

resolved

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+    private readonly  IEnumerator<Point> pointEnumerator;
+    private List<Rectangle> layoutRectangles = new List<Rectangle>();
+    
+    public CircularCloudLayouter(Point layoutCenter, double radius, double angleOffset)
```
Из того же C# 12 можно использовать [primary constructor](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#primary-constructors)

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+        pointEnumerator = new FermatSpiralPointsGenerator(radius, angleOffset)
+            .GeneratePoints(layoutCenter)
+            .GetEnumerator();
+    }
```
Было бы здорово, если `CircularCloudLayouter` мог использовать любой алгоритм генерации точек.  
Сейчас он умеет использовать только спираль Ферма

Folleach replied:
![image](https://github.com/user-attachments/assets/41a7cf18-3e16-474f-8925-2725fe28b6b5)

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+            var rectanglePos = pointEnumerator.Current;
+            rectangle = CreateRectangleWithCenter(rectanglePos, rectangleSize);
+            
+        } while (layoutRectangles.Any(rectangle.IntersectsWith));
```
Было бы здорово, если `IEnumerator` можно было продолжить более простым синтаксисом.
Тогда этот код мог бы выглядеть как-то так:
```cs
var rectangle = pointEnumerator
    ...
    .Select(x => CreateRectangleWithCenter(x, rectangleSize))
    .First(x => !layoutRectangles.Any(x.IntersectsWith));
```

Folleach replied:
Здорово!
У меня получился примерно такой же extension метод. Разве что назвал его `Continue`, потому что он продолжает перечислять `IEnumerator` с прошлого момента. `ToIEnumerable` не совсем явно это говорит. Возможно, внутри он каким-то образом клонирует `IEnumerator`?

resolved

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+        Rectangle rectangle;
+        do
+        {
+            pointEnumerator.MoveNext();
```
А если генератор будет конечный?

Folleach replied:
> Был передан конечный генератор точек

Был передан куда? В `PutNextRectangle`? Хотелось бы информативнее сообщение

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+        var x = center.X - rectangleSize.Width / 2;
+        var y = center.Y - rectangleSize.Height / 2;
+        return new Rectangle(x, y, rectangleSize.Width, rectangleSize.Height);
+    }
```
Может ли этот метод пригодиться где-нибудь ещё?

Folleach replied:
Вероятно, стоило куда-либо вынести этот метод в качестве метода расширения или что-то вроде

Folleach replied:
Странно конечно, что `extension` метод сделан над `Rectangle`, который никак не используется...

Можно было бы сделать что-то вроде
```cs
size.RectangleWithCenterAt(point)
```

resolved

Folleach commented on cs/TagsCloudVisualization/CloudLayouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,44 @@
+
+public class CircularCloudLayouter : ICircularCloudLayouter
+{
+    private readonly Point layoutCenter;
```
Не используется, но видимо будет использоваться позже

Folleach replied:
Зачем кому-то извне знать, какой центр у `layouter`?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/Visualizers/IVisualizer.cs:
```diff
@@ -0,0 +1,7 @@
+public interface IVisualizer
+{
+    public Bitmap CreateBitmap(IEnumerable<Rectangle> rectangles, Size bitmapSize);
+}
```
Я бы предложил абстрагировать интерфейс визуализации от размера полотна.
Теоретически мы можем захотеть написать реализацию, которая рисует на бесконечном полотне

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,14 @@
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <PackageReference Include="System.Drawing.Common" Version="6.0.0" />
+    </ItemGroup>
```
Почему именно 6.0.0? Есть ведь новее: [9.0.0](https://www.nuget.org/packages/System.Drawing.Common/9.0.0)

Folleach commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,47 @@
+
+    private static string GetPathToImages()
+    {
+        var filename = $"{RectanglesNumber}_{LayoutRadius}_{LayoutAngleOffset}_TagCloud.jpg";
```
Формат имени файла будет зависеть от локали компьютера, на котором запускается код. Лучше этого избегать, потому что работать с недетерминированным форматом - сложно

Folleach replied:
resolved

Но в библиотеках культуру глобально менять не стоит

Folleach commented on cs/TagsCloudVisualization/images/1000_1_5_TagCloud.jpg:
Круто! 

**Не относится к задаче**
Бинарные файлы в репозиторий лучше не загружать, или делать их маленькими на сколько возможно.
Ведь эти файлы останутся там навсегда, даже если в следующих коммитах они будут удалены.
Понимаю, что сейчас особо другого варианта не было, так что всё замечательно!

Folleach commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,47 @@
+        Directory.CreateDirectory(ImagesDirectory);
+        
+        bitmap.Save(GetPathToImages(), ImageFormat.Jpeg);
+    }
```
Возможно тут можно выделить какие-нибудь методы? И даже переиспользовать их в других местах

Folleach replied:
resolved

Folleach commented on cs/tdd.sln:
```diff
@@ -7,6 +7,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BowlingGame", "BowlingGame\
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Samples", "Samples\Samples.csproj", "{B5108E20-2ACF-4ED9-84FE-2A718050FC94}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TagsCloudVisualization", "TagsCloudVisualization\TagsCloudVisualization.csproj", "{01ADC663-2353-458D-9769-790C224690B2}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TagsCloudVisualizationTests", "TagsCloudVisualizationTests\TagsCloudVisualizationTests.csproj", "{EE74CFE7-F089-4D9E-861E-19C4D4854974}"
+EndProject
```
Ты понимаешь, что написано в этом файле?
Я тоже нет :)

В новых версиях Visual Studio и Rider можно использовать новый формат: slnx
https://blog.ndepend.com/slnx-the-new-net-solution-xml-file-format/

Он более человеко-читаем:
```xml
<Solution>
  <Folder Name="/Test/">
    <Project Path="UnitTests\UnitTests.csproj">
      <Configuration Solution="Release|*" Project="*|*|NoBuild" />
    </Project>
  </Folder>
  <Folder Name="/Application/">
    <Project Path="App\App.csproj" />
  </Folder>
</Solution>
```
Если сделаешь задачу и будет совсем скучно - можешь попробовать перевести твой sln на этот формат. А так, этот комментарий просто для информации, что вот такое существует

Folleach reviewed (changes requested):
### Резюмируя

Разделение классов сделано хорошо, мне оно понравилось.  

Недостаточно тестов, но думаю, ко второму дедлайну они будут доделаны. Простор для тестов тут, на самом деле, огромный. Не стесняйся тестировать даже то, что выходит за рамки основных классов: экстеншены, хелперы и всякие другие мелочи, которые есть.
Но как минимум, ожидаю увидеть тесты на все требования к раскладке

Также хотелось бы видеть более прибранный код: где-то улучшить декомпозицию, где-то исправить стилистические мелочи. Об этом я оставил комментарии ниже.
И не забывай обращать внимание на разнообразные предупреждения IDE, они помогают улучшить общую красоту кода ^_^

Folleach commented:
О, и ещё, такая гранулярность коммитов даёт понять, как разрабатывалась задача, но как правило, такие подробности излишни

Если уже сделал коммит и хочешь что-то добавить, используй amend.

Например коммиты
[Теперь проект структурирован по папочкам](https://github.com/kontur-courses/tdd/pull/239/commits/345306170edeac2e09665bdea7bff1f85bbd9776)
[Теперь namespace'ы в соответствии с папочками](https://github.com/kontur-courses/tdd/pull/239/commits/0d68ef1c1ff38746438ea3af4fa169cc889ae87a)

Можно было схлопнуть в один.

Или вот эти
[Добавил README.MD с картинками](https://github.com/kontur-courses/tdd/pull/239/commits/a2d65c20861a283ff0286227669e3c1de703179a)
[README.MD теперь должен нормально отображаться](https://github.com/kontur-courses/tdd/pull/239/commits/bc3bee5ea7b86d4efd63d84880bd8f06811a2f88)
[Теперь README.MD точно должен нормально отображаться](https://github.com/kontur-courses/tdd/pull/239/commits/1ff2cf21aa8238d472b15a4e88fe521ee1394312)

Folleach commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,47 @@
+        var center = new Point(ImageWidth / 2, ImageHeight / 2);
+        var cloudLayouter = new CircularCloudLayouter(center, LayoutRadius, LayoutAngleOffset);
+        var random = new Random();
+        var rectangles = new Rectangle[RectanglesNumber];
+
+        rectangles = rectangles
+            .Select(_ => cloudLayouter.PutNextRectangle(new Size(
+                random.Next(MinRectangleSize, MaxRectangleSize),
+                random.Next(MinRectangleSize, MaxRectangleSize))))
+            .ToArray();
```
Заметил позже, тут создаётся массив, чтобы заменить его другим массивом?!
Может как-то без создания `rectangles` можно вызвать `Select` нужное количество раз?

Folleach replied:
resolved

Geratoptus committed:
Добавил заготовку для RandomExtension.cs

Geratoptus committed:
Добавил красный тест для RandomExtension

Geratoptus committed:
RandomExtension.cs RandomSize теперь тест зеленый

Geratoptus committed:
RandomExtensionTest.cs добавил красный тест для RandomSize

Geratoptus committed:
RandomExtension.cs теперь тест в RandomExtensionTest.cs зеленый

Geratoptus committed:
Внедрение RandomExtension в проект

Geratoptus committed:
RandomExtension.cs
-Добавил заготовку для RandomPoint
-Добавил значения по умолчанию передаваемым параметрам

Geratoptus committed:
RandomExtensionTest.cs добавил красный тест для RandomPoint

Geratoptus committed:
RandomExtension.cs RandomPoint теперь возвращает точку (тест зеленый)

Geratoptus committed:
RandomExtensionTest.cs добавил красный тест для RandomPoint

Geratoptus committed:
RandomExtension.cs
-RandomPoint теперь возвращает ожидаемые точки
-RandomPoint исправлены значения по умолчанию

Geratoptus committed:
CircularCloudLayouterTest.cs
-Добавил тест на плотность и форму
-Поменял рандомизацию на более корректную
-Декомпозировал перегруженный тест на пересечение прямоугольнико в раскладке
-Добавил несколько более базовых тестов

Geratoptus committed:
EnumeratorExtensionTest.cs добавил красный тест для EnumeratorExtension.cs

Geratoptus committed:
EnumeratorExtension.cs ToIEnumerable теперь возвращает ожидаемую последовательность

Geratoptus committed:
IVisualizer.cs CreateBitmap теперь не закрепляет в поведении зависимость метода от bitmapSize

Geratoptus committed:
DefaultVisualizer.cs зависимость от bitmapSize вынесена в конструктор

Geratoptus committed:
CircularCloudLayouter.cs
-Внедрен EnumeratorExtension
-Теперь может работать с любой реализацией интерфейса IPointsGenerator
-Основной конструктор оформлен, как primary constructor

Geratoptus committed:
Визуальные правки

Geratoptus committed:
Program.cs
-Внедрение нового конструктора DefaultVisualizer
-Теперь имя файла не зависит от locale

Geratoptus committed:
Добавил класс DefaultBitmapSaver.cs

Geratoptus committed:
DefaultVisualizer.cs добавил оцентровку изображения

Geratoptus committed:
Program.cs внедрил DefaultBitmapSaver

Geratoptus committed:
CircularCloudLayouterTest.cs
-Добавил TearDown, в котором при падении теста сохраняется визуальная раскладка, а также информация о прямоугольниках в раскладке
-Для декомпозиции ввел вспомогательные методы

Geratoptus committed:
DefaultBitmapSaverTest.cs добавил тесты для DefaultBitmapSaver'a

Geratoptus committed:
CircularCloudLayouter.cs добавил обработку случая, если был передан конечный генератор

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,177 @@
+        if (currentContext.Result.Outcome.Status != TestStatus.Failed)
+            return;
+        
+        var visualizer = new DefaultVisualizer(GetLayoutSize(testRectangles));
```
На всякий случай я бы занулял `testRectangles` перед каждым тестом, чтобы случайно не отрисоватьт "чужие" прямоугольники

Folleach replied:
Повторять один и тот же код в каждом тесте - не удобно.  
Можно было поместить его в `SetUp`, который выполняется перед запуском каждого теста

resolved

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,177 @@
+        
+        testRectangles = [rectangle];
+        rectangle.Should().BeOfType<Rectangle>();
+    }
```
Что проверяет этот тест?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualizationTests/Extensions/EnumeratorExtensionTest.cs:
```diff
@@ -0,0 +1,29 @@
+public class EnumeratorExtensionTest
+{
+
+    [Test]
+    public void ToIEnumerable_ShouldReturnExpectedEnumerable()
+    {
+        const int elementsNumber = 1000;
+        var centerPoint = new Point(0, 0);
+        var expectedEnumerable = new FermatSpiralPointsGenerator(1, 0.5)
+            .GeneratePoints(centerPoint);
+        var actualEnumerable = expectedEnumerable
+            .GetEnumerator()
+            .ToIEnumerable();
+        
+        actualEnumerable.Take(elementsNumber).Should()
+            .BeEquivalentTo(expectedEnumerable.Take(elementsNumber));
+    }
```
Вероятно, один из важных тестов тут ещё такой: `ToIEnumerable` должен действительно продолжать перечисление, а не начинать заново

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,177 @@
+        IsIntersectionBetweenRectangles(testRectangles).Should().BeFalse();
+    }
+
+    [Test]
+    [Repeat(10)]
+    public void ShouldGenerateLayoutThatHasHighTightnessAndShapeOfCircularCloud_WhenOptimalParametersAreUsed()
+    {
+        const double allowableDelta = 0.35;
+
+        var circularCloudLayouter = SetupLayouterWithOptimalParameters(); 
+        testRectangles = PutRectanglesInLayouter(randomizer.Next(800, 1200), circularCloudLayouter);
+
+        var layoutSize = GetLayoutSize(testRectangles);
+        
+        var circumcircleDiameter = Math.Max(layoutSize.Height, layoutSize.Width);
+        var circumcircleArea = Math.PI * Math.Pow(circumcircleDiameter, 2) / 4;
+        var rectanglesArea = (double)testRectangles
+            .Select(rectangle => rectangle.Height * rectangle.Width)
+            .Sum();
+        
+        var areasFraction = circumcircleArea / rectanglesArea;
+        areasFraction.Should().BeApproximately(1, allowableDelta);        
+    }
```
У меня тут прошёл квадрат, хотя не должен вроде как...
```cs
var list = new List<Rectangle>();
for (var i = 0; i < 100; i++)
{
    for (var j = 0; j < 100; j++)
    {
        list.Add(new Rectangle(new Point(i, j), new Size(1, 1)));
    }
}
testRectangles = list.ToArray();
```

Folleach replied:
resolved

Geratoptus committed:
EnumeratorExtensionTest.cs добавлен тест, проверяющий, что ToIEnumerable продолжает перечисление

Geratoptus committed:
RectangleExtensionTest.cs добавил красные тесты для RectangleExtension.cs

Geratoptus committed:
RectangleExtension.cs теперь тесты зеленые

Geratoptus committed:
CircularCloudLayouterTest.cs
-layoutCenter запоминается при Setup'е
-ShouldGenerateLayoutThatHasHighTightnessAndShapeOfCircularCloud_WhenOptimalParametersAreUsed исправлено нахождение радиуса
-GetLayoutSize исправлено имя Enumerabl'a, с которым работает метод

Geratoptus committed:
CircularCloudLayouter.cs
-LayoutCenter убран за ненадобностью
-CreateRectangleWithCenter убран (теперь в RectangleExtension)
-PutNextRectangle вместо try-catch теперь используется FirstOrDefault + if
-PutNextRectangle сообщение в исключении теперь более информативно и вынесено в константу

Geratoptus committed:
CircularCloudLayouterTest.cs добавлен тест PutNextRectangle_ShouldThrowInvalidOperationException_IfFiniteGenerator
